snap/quota/quota.go

```diff
@@ -271,11 +272,31 @@ func (grp *Group) getAllowedCPUs() []int {
 	return grp.CPULimit.AllowedCPUs
 }
 
+// GetCorrectedCPUCount returns the maximum number of allowed CPU cores for
+// this group. It needs to take into account that CPU set might have been set
+// to limit the number of cores, or a direct limit on the number of cores.
+// Goal is to select the most restrictive limit.
+func (grp *Group) GetCorrectedCPUCount() int {
+	if grp.CPULimit == nil {
+		return 0
+	}
+
+	cpuCount := runtime.NumCPU()
```

Comment:
Are you sure that we need to take into account the real number of the CPUs? Isn't cgroup smart enough to handle the case where one specified a cpu count greater than the phisical number?

Also, what do we do with the return value of `GetCorrectedCPUCount()`? Is it only used at runtime, or are we also storing it on disk somewhere (in the latter case, this could pose a problem if CPUs are added after a snap quota has been configured, I guess)?

Reply:
Yea, systemd don't care if you writen 1000% or 1200% if there are only 4 CPUs, it'll just allow maximum. But if you create a Quota Group with 50% cpu usage, you want it to maximum use 50% of the current number of cpus. This is why we always calculate with the number of currently active cpus so we adjust for CPUs brought online at later point, or offline for that matter. This works because we never store the number of CPUs, only the percentage allowed (unless the user requests a specific cpu count). This way it doesn't matter how many CPUs are active at any given point, as all calculations work from the current core count.

We only use the return value of `GetCorrectedCPUCount()` during runtime

Reply:
Thanks, makes sense.